### PR TITLE
VxDesign: Basic election creation/editing flow

### PR DESCRIPTION
**Pull request by jonahkagan**
## Overview

Restructures the app to focus on creating/editing elections. The previous ballot proofing functionality is maintained within this flow.

_Review by commits for more details_

## Demo Video or Screenshot

Very basic walkthrough of all screens (omitting many features/edge cases)

https://github.com/votingworks/vxsuite/assets/530106/953b1bf2-23ae-4d8a-8cff-61ebd1b42a9b

## Testing Plan

Manual testing - this is still prototype UI that may change drastically

## Checklist

- [ ] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.

- [ ] I have added a screenshot and/or video to this PR to demo the change
- [ ] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

### CODE REVIEW

**Review by adghayes — APPROVED** [review]
Wow! I am impressed by your work as always, in this case by the amount of functionality you got into this prototype. I reviewed the backend changes in depth and the frontend changes more cursorily, using your video as a guide.

The precinct split approach you've applied here makes sense and I'm excited that we're able to test it out.

On the frontend, I like the major design choices and components:
- initially non-editable forms which need to be "unlocked"
- steering away from modals
- navigation bar
- tab panels
- breadcrumbs

There are a few minor design things that I would suggest iterating on if this weren't a prototype:
- the heavy use of green as both a call to action and as an app state indicator can get a little exhausting
- the `Card` component doesn't add too much, and in cases like the wrapped toggles on the contest definition page, I think it detracts in that it is unnecessary and distracting. I think the component makes the most sense in cases where you are laying out multiple like things, e.g. ballot styles
- the layout of the "Edit Precinct" page is a bit confusing in that the "Splits" end up under the "Districts" header and, although the splits _are_ being used to define different district lists, it doesn't quite make sense. when no precincts are defined, the jumble of buttons below the "Districts" header is hard to make sense of visually. I would suggest that, when there's no splits, the districts header remains with the district list below it, but the "Add Splits" button is moved either lower or higher in the page. then, when splits are added, it should be a "Splits" header with "Districts" lists or subheaders within each card. have you looked at how the app behaves if there are lots of splits defined? does app become scrollable horizontally?

Eventually I would like to see these apps be visually consistent with our other apps but that's work for much later.

QA Questions:
- Have you manually tested using an election definition heavily edited or entirely defined in this app and then loaded in VxAdmin?
- Not specific to this PR but, since this is my first dive into the app - what guarantees do we have the ballot rendered and shown to the user on the frontend is the same as the one exported to PDF on the backend? We've long tried to generate ballots in one place and it sort of looks like we're doing it in two places here, but `design/shared` means that the ballot definition and layout is shared.

Anyway, looks fantastic for the prototype, approving to unblock you for whatever are the next steps here.

**Comment by adghayes** [comment]
one more design delta I forgot in my review comments - I think the "All Elections" page needs more structure as a list, possibly showing metadata. my guess from looking at the video and the code is that with multiple elections is looks a little blobby and irregular, with all gray buttons and little to distinguish them.

again, as far as prototypes are concerned, doesn't matter

File: apps/design/backend/src/store.ts [hunk 1]
```diff
@@ -1,10 +1,172 @@
...
+  name: string;
+  districtIds: DistrictId[];
+}
+export interface PrecinctWithSplits {
+  id: PrecinctId;
+  name: string;
+  splits: PrecinctSplit[];
+}
+export interface PrecinctSplit {
+  id: PrecinctId;
```
[the remark this answers is not in the document]
Comment (jonahkagan) [reply apps/design/backend/src/store.ts:37]: Great point!

File: apps/design/backend/src/store.ts [hunk 2]
```diff
@@ -1,10 +1,172 @@
...
+  name: string;
+  districtIds: DistrictId[];
+}
+export type Precinct = PrecinctWithoutSplits | PrecinctWithSplits;
+
+export function hasSplits(precinct: Precinct): precinct is PrecinctWithSplits {
+  return 'splits' in precinct && precinct.splits !== undefined;
+}
+
+// We also create a new type for a ballot style, that can reference precints and
```
[the remark this answers is not in the document]
Comment (jonahkagan) [reply apps/design/backend/src/store.ts:47]: Literally the word I misspell the most often other than jurisdiction

(I found spell check too annoying for code so I turned it off)

File: apps/design/backend/src/store.ts [hunk 3]
```diff
@@ -1,10 +1,172 @@
...
+      })),
+    };
+  });
+}
+
+function groupBy<T, K>(items: T[], keyFn: (item: T) => K): Array<[K, T[]]> {
+  const groups: Array<[K, T[]]> = [];
+  for (const item of items) {
+    const key = keyFn(item);
+    const group = groups.find(([k]) => deepEqual(k, key));
```
[the remark this answers is not in the document]
Comment (jonahkagan) [reply apps/design/backend/src/store.ts:86]: Will do

File: apps/design/backend/src/store.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (eventualbuddha) [reply apps/design/backend/src/store.ts:47]: If there are words that we use that it's flagging that you think it shouldn't we can add them to the dictionary.

[the remark this answers is not in the document]
Comment (adghayes) [reply apps/design/backend/src/store.ts:47]: something about "precincts" is distinctly difficult to type, I relate

**Comment by jonahkagan** [comment]
> the heavy use of green as both a call to action and as an app state indicator can get a little exhausting
Agreed, I think this is a limitation of the current design system. Will try to work around.

> the Card component doesn't add too much, and in cases like the wrapped toggles on the contest definition page, I think it detracts in that it is unnecessary and distracting. I think the component makes the most sense in cases where you are laying out multiple like things, e.g. ballot styles
The border around the toggles is actually a part of the existing SegmentedButton component - it's not a Card. But if there are other places you think I could remove the Card let me know.

> the layout of the "Edit Precinct" page is a bit confusing in that the "Splits" end up under the "Districts" header
Yeah, this was a bit tricky. I think I will leave this for now and revisit in a later phase.

> have you looked at how the app behaves if there are lots of splits defined? does app become scrollable horizontally?
The cards wrap to a new line (using flex-wrap)

>  I think the "All Elections" page needs more structure as a list, possibly showing metadata
Agreed! Will revisit in a later phase.

**Pull request merged**